app/code/Magento/CatalogUrlRewrite/Observer/ProcessUrlRewriteOnChangeProductVisibilityObserver.php

```diff
@@ -0,0 +1,111 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\CatalogUrlRewrite\Observer;
+
+use Magento\Catalog\Api\Data\ProductInterface;
+use Magento\Catalog\Model\Product\Visibility;
+use Magento\CatalogUrlRewrite\Model\ProductUrlPathGenerator;
+use Magento\CatalogUrlRewrite\Model\ProductUrlRewriteGenerator;
+use Magento\UrlRewrite\Model\Exception\UrlAlreadyExistsException;
+use Magento\UrlRewrite\Model\UrlPersistInterface;
+use Magento\UrlRewrite\Service\V1\Data\UrlRewrite;
+use Magento\Catalog\Model\ResourceModel\Product\CollectionFactory;
+use Magento\Framework\Event\Observer;
+use Magento\Framework\Event\ObserverInterface;
+
+/**
+ * Class ProductProcessUrlRewriteSavingObserver
+ */
+class ProcessUrlRewriteOnChangeProductVisibilityObserver implements ObserverInterface
+{
+    /**
+     * @var CollectionFactory
+     */
+    private $productCollectionFactory;
+
+    /**
+     * @var ProductUrlRewriteGenerator
+     */
+    private $urlRewriteGenerator;
+
+    /**
+     * @var UrlPersistInterface
+     */
+    private $urlPersist;
+
+    /**
+     * @var ProductUrlPathGenerator
+     */
+    private $urlPathGenerator;
+
+    /**
+     * @param CollectionFactory $collectionFactory
+     * @param ProductUrlRewriteGenerator $urlRewriteGenerator
+     * @param UrlPersistInterface $urlPersist
+     * @param ProductUrlPathGenerator|null $urlPathGenerator
+     */
+    public function __construct(
+        CollectionFactory $collectionFactory,
+        ProductUrlRewriteGenerator $urlRewriteGenerator,
+        UrlPersistInterface $urlPersist,
+        ProductUrlPathGenerator $urlPathGenerator
+    ) {
+        $this->productCollectionFactory = $collectionFactory;
+        $this->urlRewriteGenerator = $urlRewriteGenerator;
+        $this->urlPersist = $urlPersist;
+        $this->urlPathGenerator = $urlPathGenerator;
+    }
+
+    /**
+     * Generate urls for UrlRewrites and save it in storage
+     *
+     * @param Observer $observer
+     * @return array
+     * @throws UrlAlreadyExistsException
+     */
+    public function execute(Observer $observer)
+    {
+        $event = $observer->getEvent();
+        $attrData = $event->getAttributesData();
+        $productIds = $event->getProductIds();
+        $storeId = $event->getStoreId();
+        $visibility = $attrData[ProductInterface::VISIBILITY] ?? null;
+
+        if (!$visibility) {
```

Comment:
@VitaliyBoyko I don't really get this logic. URL Rewrites for invisible products should still be created so that they can be made visible at any moment without conflict.

Please update description with current/changed behavior.

Reply:
Hi @orlangur 
I've updated the PR description. Hope it is more comprehensive now. You may also check 
https://github.com/magento/magento2/issues/10725
https://github.com/magento/magento2/issues/20434 for additional information.
Thank you!